### PR TITLE
Unsupported content type

### PR DESCRIPTION
**Pull request by 82marbag**
## Motivation and Context

Add validation for the Content-Type header and pass (remove from the failing list) the relevant protocol tests.

Closes: https://github.com/awslabs/smithy-rs/issues/1663

## Testing

`./gradlew :codegen-server-test:test && ./gradlew :codegen-test:test`

## Checklist

- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the smithy-rs codegen or runtime crates
- [ ] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:731]: I think we should split this code. There should be two functions, each handling the case of whether the body is empty or not.

**Review by david-perez — CHANGES_REQUESTED** [review]

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (82marbag) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:189]: Thanks. The `break` in the loop was doing the same, but this is cleaner

File: rust-runtime/aws-smithy-http-server/src/protocols.rs
Comment (82marbag) [reply rust-runtime/aws-smithy-http-server/src/protocols.rs:36]: No, it's just that a client should always send an empty JSON object payload in AwsJson 1.0, so with `{}` in the body this would fail (this only checks whether the body is empty)

File: rust-runtime/aws-smithy-http-server/src/protocols.rs
Comment (82marbag) [reply rust-runtime/aws-smithy-http-server/src/protocols.rs:57]: Added docs explaining what they do, but neither panic

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (82marbag) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:731]: But when the body is not empty, it's just `Ok(())`

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (82marbag) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:700]: This is on purpose because of [this test](https://github.com/awslabs/smithy/blob/f8968b9462db03e858b20938c7e82c0c5ac39ac0/smithy-aws-protocol-tests/model/restJson1/empty-input-output.smithy#L199). It's the first example you give, but has a non-empty body and a content-type

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/Protocol.kt
Comment (jdisanti) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/Protocol.kt:99]: I don't want to block approval on this, but now that the prerequisite work has been done in #1698, it would be better to have this in the `ServerProtocol` implementations instead.

**Review by jdisanti — APPROVED** [review]

File: rust-runtime/aws-smithy-http-server/src/protocols.rs
Comment (david-perez) [inline rust-runtime/aws-smithy-http-server/src/protocols.rs:82]:
```suggestion
            // `expected_content_type` comes from the codegen.
```

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/Protocol.kt
Comment (david-perez) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/Protocol.kt:97]:
```suggestion
     * In some protocols, such as restJson1,
     * when there is no modeled body input, content type must not be set and the body must be empty.
     * Returns a boolean indicating whether to perform this check.
```

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (david-perez) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:730]: Clippy does not flag this `if true`?

I'd rewrite this to avoid it. You can use `conditionalBlock("if body.is_empty() {", "}", conditional = parser != null) { ... }` for this purpose.

**Review by david-perez — APPROVED** [review]

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (david-perez) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:700]: Ah I see. Interesting. What's the protocol test that asserts that body must be empty when there is no modeled input? I can't seem to find it right now.

Comment (82marbag) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:700]: [This one](https://github.com/awslabs/smithy/blob/603c0f2de0dc54c7ed5cfa0d1e2f1c63d521eed2/smithy-aws-protocol-tests/model/restJson1/malformedRequests/malformed-content-type.smithy#L12) (which corresponds to the second example you give)

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt
Comment (82marbag) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt:730]: Thanks!

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/Protocol.kt
Comment (hlbarber) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/Protocol.kt:99]: After this is merged I'll move this in https://github.com/awslabs/smithy-rs/pull/1731

**Pull request merged**
